Fix VSTS 669803: using vstool with Razor

KirillOsenkov opened the pull request:
vstool/mdtool was not setting the Runtime.MainSynchronizationContext, which in turn didn't set the MainTaskScheduler. Razor addin relies on the MainTaskScheduler being set. Repro: vstool build Foo.sln failed with ArgumentNullException.

KirillOsenkov committed:
Fix VSTS 669803: using vstool with Razor

vstool/mdtool was not setting the Runtime.MainSynchronizationContext, which in turn didn't set the MainTaskScheduler. Razor addin relies on the MainTaskScheduler being set. Repro: vstool build Foo.sln failed with ArgumentNullException.

KirillOsenkov commented:
@monojenkins backport release-7.6

KirillOsenkov commented:
@monojenkins backport release-7.7

Therzok commented:
The reason I didn't do this change initially is that other vstool tools set their own MainTaskScheduler - i.e. BuildTool. Those need to be removed, as setting it will throw.

Therzok commented:
And if run-md-tests set it, other applications might.

KirillOsenkov committed:
Allow re-setting the MainSynchronizationContext with a warning.

This is needed because mdtool sets it, and then TestRunner needs to re-set it to an XwtSynchronizationContext.

KirillOsenkov commented:
See also: https://github.com/mono/monodevelop/pull/4863

KirillOsenkov commented:
@monojenkins backport release-7.7
